sqlmesh/core/context.py

```diff
@@ -1255,27 +1259,45 @@ def _snapshots(
                         if name not in audits:
                             audits[name] = audit
 
-        snapshots = {}
+        def _nodes_to_snapshots(nodes: t.Dict[str, Node]) -> t.Dict[str, Snapshot]:
+            snapshots: t.Dict[str, Snapshot] = {}
+            fingerprint_cache: t.Dict[str, SnapshotFingerprint] = {}
 
-        for node in nodes.values():
-            if node.name not in local_nodes and node.name in remote_snapshots:
-                snapshot = remote_snapshots[node.name]
-                ttl = snapshot.ttl
-            else:
-                config = self.config_for_node(node)
-                ttl = config.snapshot_ttl
-
-            snapshot = Snapshot.from_node(
-                node,
-                nodes=nodes,
-                audits=audits,
-                cache=fingerprint_cache,
-                ttl=ttl,
-            )
-            snapshots[node.name] = snapshot
+            for node in nodes.values():
+                if node.name not in local_nodes and node.name in remote_snapshots:
+                    snapshot = remote_snapshots[node.name]
+                    ttl = snapshot.ttl
+                else:
+                    config = self.config_for_node(node)
+                    ttl = config.snapshot_ttl
+
+                snapshot = Snapshot.from_node(
+                    node,
+                    nodes=nodes,
+                    audits=audits,
+                    cache=fingerprint_cache,
+                    ttl=ttl,
+                )
+                snapshots[node.name] = snapshot
+            return snapshots
 
+        snapshots = _nodes_to_snapshots(nodes)
         stored_snapshots = self.state_reader.get_snapshots(snapshots.values())
 
+        non_revertible_snapshot_names = {
+            s.name for s in stored_snapshots.values() if s.name in local_nodes and s.non_revertible
+        }
+        if non_revertible_snapshot_names:
+            for name in non_revertible_snapshot_names:
+                snapshot_id = snapshots[name].snapshot_id
+                logger.info(
+                    "Found a non-revertible snapshot %s. Restamping the model...", snapshot_id
+                )
+                node = local_nodes[name]
+                nodes[name] = node.copy(update={"stamp": f"revert to {snapshot_id.identifier}"})
```

Comment:
what does identifier str() turn into?

Reply:
`identifier` is a string

Reply:
it's a hash string

Reply:
perfect, wanted to make sure it's not like Identifier(id: ...)